pkg/controller/rollout/rollout_bluegreen.go

```diff
@@ -0,0 +1,487 @@
+		c.NewStatus.Message = cond.Message
+		return false, nil
+	}
+	cond.Message = fmt.Sprintf("Rollout is in step(%d/%d), and wait duration(%d seconds) to enter the next step", blueGreenStatus.CurrentStepIndex, steps, *currentStep.Pause.Duration)
+	c.NewStatus.Message = cond.Message
+	// wait duration time, then go to next step
+	duration := time.Second * time.Duration(*currentStep.Pause.Duration)
+	expectedTime := blueGreenStatus.LastUpdateTime.Add(duration)
+	if expectedTime.Before(time.Now()) {
+		klog.Infof("rollout(%s/%s) canary step(%d) paused duration(%d seconds), and go to the next step",
+			c.Rollout.Namespace, c.Rollout.Name, blueGreenStatus.CurrentStepIndex, *currentStep.Pause.Duration)
+		return true, nil
+	}
+	c.RecheckTime = &expectedTime
+	return false, nil
+}
+
+func (m *blueGreenReleaseManager) doCanaryJump(c *RolloutContext) (jumped bool) {
+	bluegreenStatus := c.NewStatus.BlueGreenStatus
+	nextIndex := bluegreenStatus.NextStepIndex
+	/*
+		we set the CurrentStepIndex same as NextStepIndex to prevent currentStepIndex from out of range
+		for example, if we had a rollout with 4 steps and CurrentStepIndex was 2
+		then, the user removed 3 steps from the plan, we can calculate NextStepIndex is 1 correctly,
+		but CurrentStepIndex remains 2, which could cause out of range.
+	*/
+	resetCurrentIndex := false
+	if int(bluegreenStatus.CurrentStepIndex) > len(c.Rollout.Spec.Strategy.BlueGreen.Steps) {
+		bluegreenStatus.CurrentStepIndex = nextIndex
+		resetCurrentIndex = true
+	}
+	currentStep := c.Rollout.Spec.Strategy.BlueGreen.Steps[bluegreenStatus.CurrentStepIndex-1]
+	if resetCurrentIndex || nextIndex != util.NextBatchIndex(c.Rollout, bluegreenStatus.CurrentStepIndex) && nextIndex > 0 {
+		currentIndexBackup := bluegreenStatus.CurrentStepIndex
+		currentStepStateBackup := bluegreenStatus.CurrentStepState
+		bluegreenStatus.CurrentStepIndex = nextIndex
+		bluegreenStatus.NextStepIndex = util.NextBatchIndex(c.Rollout, nextIndex)
+		nextStep := c.Rollout.Spec.Strategy.BlueGreen.Steps[nextIndex-1]
+		// if the Replicas between currentStep and nextStep is same, we can jump to
+		// the TrafficRouting step; otherwise, we should start from the Init step
+		if reflect.DeepEqual(nextStep.Replicas, currentStep.Replicas) && !resetCurrentIndex {
+			bluegreenStatus.LastUpdateTime = &metav1.Time{Time: time.Now()}
+			bluegreenStatus.CurrentStepState = v1beta1.CanaryStepStateTrafficRouting
+			klog.Infof("rollout(%s/%s) step(%d) state from(%s) -> to(%s)", c.Rollout.Namespace, c.Rollout.Name,
+				bluegreenStatus.CurrentStepIndex, currentStepStateBackup, bluegreenStatus.CurrentStepState)
+		} else {
+			bluegreenStatus.LastUpdateTime = &metav1.Time{Time: time.Now()}
+			bluegreenStatus.CurrentStepState = v1beta1.CanaryStepStateInit
+			klog.Infof("rollout(%s/%s) step(%d) state from(%s) -> to(%s)", c.Rollout.Namespace, c.Rollout.Name,
```

Comment:
it's a little hard to do so, here is why
1. only if a step is done, can we run `bluegreenStatus.LastUpdateTime = &metav1.Time{Time: time.Now()}`, if not done, we cannot update the `bluegreenStatus.LastUpdateTime`. 
2. klog Infof is reasonable to not shared by every if conditions either, it can help us to realize the fallthrough, what's more, if we use a new variable currentStepStateBackup additionally, the total lines seems keep the same....